src/Parsers/ParserSetQuery.cpp

```diff
@@ -89,6 +74,116 @@ class ParserLiteralOrMap : public IParserBase
+{
+    ParserLiteral literal_p;
+    ParserArrayOfLiterals array_p;
+    ParserTupleOfLiterals tuple_p;
+    ParserCompoundIdentifier identifier_p;
+    ParserToken s_eq(TokenType::Equals);
+
+    ASTPtr name, value;
+    String name_str, value_str;
+
+    if (!identifier_p.parse(pos, name, expected))
+        return false;
+
+    if (!s_eq.ignore(pos, expected))
+        return false;
+
+    tryGetIdentifierNameInto(name, name_str);
+
+    if (!name_str.starts_with(QUERY_PARAMETER_NAME_PREFIX))
+        return false;
+
+    name_str = name_str.substr(strlen(QUERY_PARAMETER_NAME_PREFIX));
+    if (name_str.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parameter name cannot be empty");
+
+    if (literal_p.parse(pos, value, expected)
+        || array_p.parse(pos, value, expected)
+        || tuple_p.parse(pos, value, expected))
+    {
+
+        value_str = applyVisitor(FieldVisitorToString(), value->as<ASTLiteral>()->value);
+
+        /// writeQuoted is not always quoted in line with SQL standard https://github.com/ClickHouse/ClickHouse/blob/master/src/IO/WriteHelpers.h
+        if (value_str.starts_with('\''))
+        {
+            ReadBufferFromOwnString buf(value_str);
+            readQuoted(value_str, buf);
+        }
+    }
+    else if (identifier_p.parse(pos, value, expected))
+    {
+        tryGetIdentifierNameInto(value, value_str);
+    }
+    else
+    {
+        ParserToken l_br_p(TokenType::OpeningCurlyBrace);
+        ParserToken r_br_p(TokenType::ClosingCurlyBrace);
+        ParserToken comma_p(TokenType::Comma);
+        ParserToken colon_p(TokenType::Colon);
+
+        if (!l_br_p.ignore(pos, expected))
+            return false;
+
+        int depth = 1;
+
+        value_str = '{';
+
+        while (depth > 0)
+        {
+            if (r_br_p.ignore(pos, expected))
+            {
+                value_str += '}';
+                --depth;
+                continue;
+            }
+
+            if (value_str.back() != '{')
+            {
+                if (!comma_p.ignore(pos, expected))
+                    return false;
+
+                value_str += ',';
+            }
+
+            ASTPtr key;
+            ASTPtr val;
+
+            if (!literal_p.parse(pos, key, expected))
+                return false;
+
+            if (!colon_p.ignore(pos, expected))
+                return false;
+
+            value_str += applyVisitor(FieldVisitorToString(), key->as<ASTLiteral>()->value);
+            value_str += ":";
+
+            if (l_br_p.ignore(pos, expected))
+            {
+                value_str += '{';
+                ++depth;
+                continue;
+            }
+
+            if (!literal_p.parse(pos, val, expected)
+                && !array_p.parse(pos, val, expected)
+                && !tuple_p.parse(pos, val, expected))
+                return false;
+
+            value_str += applyVisitor(FieldVisitorToString(), val->as<ASTLiteral>()->value);
```

Comment:
> We can add a separate visitor though for this task if it looks like a better way to do it.

sounds good, if it will allow to reuse the code of ParserLiteralOrMap instead of adding similar code.